### PR TITLE
gh-98762: Fix locations of match sub-patterns

### PR DESCRIPTION
**Pull request by iritkatriel**
This PR removes the recursive updates which made the locations incorrect. 

They are still too broad in some cases, but we should deal with the separately. (One thing we would need to do is to save the locations of identifiers alongside the "stores" list.)

* Issue: gh-98762

### CODE REVIEW

File: Misc/NEWS.d/next/Core and Builtins/2022-10-27-16-42-16.gh-issue-98762.Eb2kzg.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix source locations of match sub-expressions.
```
Comment (iritkatriel) [inline Misc/NEWS.d/next/Core and Builtins/2022-10-27-16-42-16.gh-issue-98762.Eb2kzg.rst]:
```suggestion
Fix source locations of :keyword:`match` sub-expressions.
```

**Review by markshannon — COMMENTED** [review]
Looks fine to me, but I'll leave it to @brandtbucher to review as he wrote the pattern matching code.

File: Lib/test/test_compile.py [hunk 2]
```diff
@@ -1130,7 +1130,8 @@ def check_positions_against_ast(self, snippet):
         class SourceOffsetVisitor(ast.NodeVisitor):
             def generic_visit(self, node):
                 super().generic_visit(node)
-                if not isinstance(node, ast.expr) and not isinstance(node, ast.stmt):
+                skip = [ast.expr, ast.stmt, ast.pattern]
+                if all(not isinstance(node, s) for s in skip):
```
Comment (brandtbucher) [inline Lib/test/test_compile.py]: This is the same, right? I find it much less confusing to follow:

```suggestion
                if not isinstance(node, (ast.expr, ast.stmt, ast.pattern)):
```

File: Misc/NEWS.d/next/Core and Builtins/2022-10-27-16-42-16.gh-issue-98762.Eb2kzg.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Fix source locations of :keyword:`match` sub-expressions.
```
Comment (brandtbucher) [inline Misc/NEWS.d/next/Core and Builtins/2022-10-27-16-42-16.gh-issue-98762.Eb2kzg.rst]:
```suggestion
Fix source locations of :keyword:`match` sub-patterns.
```

**Review by brandtbucher — APPROVED** [review]
Thanks for tackling this!

**Pull request merged**